### PR TITLE
🌱 Remove RequeueAfterError from CABPK reconcile discovery

### PR DESCRIPTION
**Pull request by vincepri**
Signed-off-by: Vince Prignano <vincepri@vmware.com>

**What this PR does / why we need it**:
Removes another use of RequeueAfterError from our codebase

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Related to #3370

/assign @ncdc

### CODE REVIEW

File: bootstrap/kubeadm/controllers/kubeadmconfig_controller.go [hunk 1]
```diff
@@ -441,13 +440,9 @@ func (r *KubeadmConfigReconciler) joinWorker(ctx context.Context, scope *Scope)
...
-	if err := r.reconcileDiscovery(ctx, scope.Cluster, scope.Config, certificates); err != nil {
-		if requeueErr, ok := errors.Cause(err).(capierrors.HasRequeueAfterError); ok {
-			scope.Info(err.Error())
-			return ctrl.Result{RequeueAfter: requeueErr.GetRequeueAfter()}, nil
-		}
-		return ctrl.Result{}, err
+	// Ensure that joinConfiguration.Discovery is properly set for joining node on the current cluster.
+	if res, err := r.reconcileDiscovery(ctx, scope.Cluster, scope.Config, certificates); err != nil || !res.IsZero() {
+		return res, err
 	}
```
Comment (detiber) [inline bootstrap/kubeadm/controllers/kubeadmconfig_controller.go:446]: This feels a bit odd given the convention of not relying on other values when an error occurs, might be better to be explicit here:
```suggestion
	res, err := r.reconcileDiscovery(ctx, scope.Cluster, scope.Config, certificates)
	if err != nil  {
		return ctrl.Result{}, err
	}
	if !res.IsZero() {
		return res, nil
	}
```

File: bootstrap/kubeadm/controllers/kubeadmconfig_controller.go [hunk 2]
```diff
@@ -524,13 +519,9 @@ func (r *KubeadmConfigReconciler) joinControlplane(ctx context.Context, scope *S
...
```
Comment (detiber) [inline bootstrap/kubeadm/controllers/kubeadmconfig_controller.go:525]: This feels a bit odd given the convention of not relying on other values when an error occurs, might be better to be explicit here:
```suggestion
	res, err := r.reconcileDiscovery(ctx, scope.Cluster, scope.Config, certificates)
	if err != nil  {
		return ctrl.Result{}, err
	}
	if !res.IsZero() {
		return res, nil
	}
```

File: bootstrap/kubeadm/controllers/kubeadmconfig_controller.go (hunk 1 above)
Comment (vincepri) [reply bootstrap/kubeadm/controllers/kubeadmconfig_controller.go:446]: I like that!

Comment (vincepri) [reply bootstrap/kubeadm/controllers/kubeadmconfig_controller.go]: Fixed it up, although I kept the block with an else to keep the variables scoped, wdyt?

Comment (detiber) [reply bootstrap/kubeadm/controllers/kubeadmconfig_controller.go]: Works for me, @ncdc is the one that despises `else` :joy:

File: bootstrap/kubeadm/controllers/kubeadmconfig_controller.go [hunk 3]
```diff
@@ -709,7 +704,7 @@ func (r *KubeadmConfigReconciler) reconcileDiscovery(ctx context.Context, cluste
 		hashes, err := certificates.GetByPurpose(secret.ClusterCA).Hashes()
 		if err != nil {
 			log.Error(err, "Unable to generate Cluster CA certificate hashes")
-			return err
+			return ctrl.Result{}, err
```
Comment (fabriziopandini) [inline bootstrap/kubeadm/controllers/kubeadmconfig_controller.go:707]: Wrap error to provide more context?

File: bootstrap/kubeadm/controllers/kubeadmconfig_controller.go [hunk 4]
```diff
@@ -730,12 +726,12 @@ func (r *KubeadmConfigReconciler) reconcileDiscovery(ctx context.Context, cluste
 	if config.Spec.JoinConfiguration.Discovery.BootstrapToken.Token == "" {
 		remoteClient, err := r.remoteClientGetter(ctx, r.Client, util.ObjectKey(cluster), r.scheme)
 		if err != nil {
-			return err
+			return ctrl.Result{}, err
```
Comment (fabriziopandini) [inline bootstrap/kubeadm/controllers/kubeadmconfig_controller.go:729]: Wrap error?

File: bootstrap/kubeadm/controllers/kubeadmconfig_controller.go (hunk 3 above)
Comment (vincepri) [reply bootstrap/kubeadm/controllers/kubeadmconfig_controller.go:707]: The error seems it's already wrapped in the `.Hashes()` method

File: bootstrap/kubeadm/controllers/kubeadmconfig_controller.go (hunk 4 above)
Comment (vincepri) [reply bootstrap/kubeadm/controllers/kubeadmconfig_controller.go:729]: Same as above, the returned error is already wrapped

File: bootstrap/kubeadm/controllers/kubeadmconfig_controller.go (hunk 3 above)
Comment (fabriziopandini) [reply bootstrap/kubeadm/controllers/kubeadmconfig_controller.go:707]: thanks for double checking

**Pull request merged**